### PR TITLE
Rework Peatio accounting using double entry accounting system 

### PR DESCRIPTION
**Pull request by ysv**
- [x] Add Operation Base model

- [x] Add Operations::Asset,Liability,Revenue,Expense

- [x] Add Operations::Chart which define class of items for which money spent or received 

- [x] Record Asset, Liability and Revenue operations on Deposit accept

- [x] Record Liability transfer on Order create

- [x] Record Liability transfer on Order cancel

- [x] Record Liability transfer and Revenue creation on Trade create

- [x] Deprecate legacy funds plus lock unlock and subtract funds

- [x] Add Admin rubric for tracking Operations and Balance sheet

- [x] Add Admin with Balance sheet

### CODE REVIEW

File: app/models/account.rb [hunk 1]
```diff
@@ -38,102 +37,97 @@ def payment_address!
     end
   end
 
-  def plus_funds!(amount)
```
Comment (ysv) [inline app/models/account.rb]: This methods should be removed

File: app/models/account.rb [hunk 2]
```diff
@@ -2,19 +2,18 @@
...
 
   ZERO = 0.to_d
 
+  has_many :operations
   has_many :payment_addresses, -> { order(id: :asc) }
 
-  validates :member_id, uniqueness: { scope: :currency_id }
-  validates :balance, :locked, numericality: { greater_than_or_equal_to: 0.to_d }
+  validates :member_id, uniqueness: { scope: %i[currency_id code] }
+  # validates :balance, :locked, numericality: { greater_than_or_equal_to: 0.to_d }
```
Comment (ysv) [inline app/models/account.rb]: Remove this line

File: app/services/accounting_service/member_entry.rb [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    def main_account; account(kind: :main); end
+    def locked_account; account(kind: :locked); end
+    memoize :main_account, :locked_account
+
+    def with_balance_check!
+      ActiveRecord::Base.transaction do
+        yield
+        # TODO: Custom Exception message.
+        # TODO: AccountingService::Error.
+        raise Account::AccountError, "Cannot create operation" if balance < 0
```
Comment (ysv) [inline app/services/accounting_service/member_entry.rb]: User friendly error message
`AccountingService::Error`

File: app/trading/matching/executor.rb [hunk 4]
```diff
@@ -145,8 +145,12 @@ def strike(trade, order, outcome_account, income_account)
       fee                         = income_value * order.fee
       real_income_value           = income_value - fee
 
-      outcome_account.assign_attributes outcome_account.attributes_after_unlock_and_sub_funds!(outcome_value)
-      income_account.assign_attributes income_account.attributes_after_plus_funds!(real_income_value)
+      # outcome_account.assign_attributes outcome_account.attributes_after_unlock_and_sub_funds!(outcome_value)
```
Comment (ysv) [inline app/trading/matching/executor.rb]: Remove this after testing

File: spec/factories/account.rb [hunk 5]
```diff
@@ -5,9 +5,22 @@ module AccountFactory
...
+      ).accept!
     end
+    if attributes.key?(:locked)
+      create(
+        "#{currency}_withdraw",
+        member: member,
+        sum: attributes.fetch(:locked)
+      ).submit!
+    end
+    member.ac(currency)
```
Comment (ysv) [inline spec/factories/account.rb]: This returns AccoutingServcie::*Entry but should be Account

File: spec/factories/withdraw.rb [hunk 6]
```diff
@@ -8,13 +8,6 @@
     rid { Faker::Bitcoin.address }
     sum { 10.to_d }
     type 'Withdraws::Coin'
-
-    account do
```
Comment (ysv) [inline spec/factories/withdraw.rb]: Number updates in DB will not work now 
All operations need to have proofs 

Don't need to update balance user is expected to have enough funds for withdrawal

File: spec/models/account_spec.rb [hunk 7]
```diff
@@ -2,57 +2,10 @@
 # frozen_string_literal: true
 
 describe Account do
-  subject { create_account(:btc, balance: '10.0'.to_d, locked: '10.0'.to_d) }
+  let(:initial_balance) { 10.0 }
```
Comment (ysv) [inline spec/models/account_spec.rb]: Account specs need to be simple
1.  association specs 
2. validations specs
3. No business logic in model (same for specs)

File: spec/services/accounting_service/member_entry_spec.rb [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+  let(:currency_id) { :btc }
+
+  let(:account) { create_account(currency_id, balance: initial_balance, locked: initial_locked) }
+  subject { AccountingService.find_or_create_for(account.member, currency_id) }
+
+  it 'sums locked and balance' do
+    expect(subject.amount).to be_d initial_balance + initial_locked
+  end
+
+  context 'operation amount less than account balance' do
```
Comment (ysv) [inline spec/services/accounting_service/member_entry_spec.rb]: Add more specs. Missing contexts are:
1. 'operation amount is negative/zero'
2. 'operation is greater than balance/locked'
3. 'concurrent operations'
4. 'double operations'

**Comment by martianatwork** [comment]
Hey this looks good.
Can we add a token id to every transaction? This will be helpful for referencing the transaction.

Instead of have credit debit columns we can have amount and type where  type can be credit or debit

**Comment by ysv** [comment]
@martianatwork we have reference for each operations I think it the same as token

Credit ant debit are commonly used for such kind of tables

File: app/models/operation.rb [hunk 9]
```diff
@@ -52,10 +52,12 @@ def debit!(reference:, amount:, kind: :main, currency: nil)
...
       if currency.blank?
-        defaul_balance = Hash[Currency.pluck(:id).collect { |id| [id.to_sym, 0] }]
-        group(:currency_id)
-        .sum('credit - debit')
-        .merge defaul_balance
+        db_balances = group(:currency_id)
+                        .sum('credit - debit')
+
+        Currency.ids.map(&:to_sym).each_with_object({}) do |id, memo|
+          memo[id] = db_balances[id] || 0
```
Comment (dinesh-skyach) [inline app/models/operation.rb]: @ysv `db_balances[id.to_s]`

**Comment by ysv** [comment]
- [x] Specs for Order

- [x] Validate operations system using [toolbox](https://github.com/rubykube/toolbox) and [validating script](https://gist.github.com/ysv/2913d720bc366ee14d0a997960556d75)
  - Single market two traders no fees
  - Multiple market 5 traders no fees
  - Single market with fees two traders
  - Multiple market with fees 5 traders
- [x] Specs for Trade

- [ ] Complex specs

- [ ] Code refactor


<s>Operation table: 
* account has_many operations
* operations belongs_to account
* account_id as a foreign key

TODO:
- [ ] Fix specs
- [ ] Test & fix legacy ui
- [ ] Test & fix admin pannel
- [ ] Prepare migrations for creating operations after removing of balance and locked
- [ ] Move #payment_address from Account to AccountingService::MemberEntry
- [ ] AccountingService specs
- [ ] AccountingService::Chart specs

Platform Accounting story
Platform owner is considered to be user with id 1
- [ ] Design & Implement Accounting Chart for Platform owner
- [ ] Implement AccountingService::PlatformEntry
- [ ] Check existing code to make double operations on withdraw & deposit (eg. deposit causes credit for liabilities & assets accounts)
</s>

**Pull request merged**